Add controller model and tests

ArthurAtangana opened the pull request:
## PR type

- [x] Feature
- [ ] Bug Fix
- [ ] Refactor
- [ ] Performance Optimization
- [ ] Documentation Update

## Description
<!--
A description of what this PR is, not how it does what it is supposed to do.
-->

## Related Issues/Tickets
<!--
If this PR is related to a ticket/issue, include it here.
For github issues the format is as follows: "closes #1234" which would automatically close the github issue
-->
- Related Issue #
- Closes #
## QA Instruction, Screenshots etc...
<!--
Information on how to run/test the changes in the PR.
Include expected/actual results here as well.
-->

## Added/updated tests?

- [x] Yes
- [ ] No: _replace this line with details on why tests are not included in this PR_

## PR Checklist

- [x] I have rebased this branch against main
- [x] All the tests are passing
- [x] My commit messages are descriptive and usefull

ArthurAtangana committed:
Resolve conflicts in cards.hpp

orselane commented on main/include/src/shared_data/commands.hpp:
```diff
@@ -33,5 +33,22 @@ inline std::ostream& operator<<(std::ostream& out, const decision& command) {
 
+inline std::ostream& operator<<(std::ostream& out, const outcome& o){
+    switch (o) {
+        case outcome::CONTINUE: out << "CONTINUE"; break;
```
? why have continue

orselane commented on main/include/test/test_couplings/controller_test.hpp:
```diff
@@ -0,0 +1,22 @@
+        //Coupling
+        addCoupling(mockGameModel->startPort, controllerModel->startInPort);
+        addCoupling(mockGameModel->decisionOutPort, controllerModel->decisionInPort);
+        addCoupling(mockGameModel->cardValueOutPort, controllerModel->valueInPort);
```
this should actually be hand value, but its just a name. As long as its semantically doing that we're fine

orselane replied:
looking at mock_game, I think its good

ArthurAtangana committed:
resolve rebase conflicts in main.cpp

orselane commented on main/include/src/atomic_models/controller.hpp:
```diff
@@ -0,0 +1,203 @@
+        }
+        else if (state.state == controllerActions::HIT) {
+            if (state.current_player == Players::CHALLENGER){
+                commandOutPort->addMessage(deckCommand::DRAW_CHALLENGER);
```
Should be two out ports, one for player, one for draw. Deck then takes draw, player as 2 inputs as well. Ends up removing the need for the nested if/else.

ArthurAtangana replied:
two ports in controller and deck for commands and player is added

orselane commented on main/include/src/atomic_models/controller.hpp:
```diff
@@ -0,0 +1,203 @@
+                commandOutPort->addMessage(deckCommand::DRAW_DEALER);
+            }
+        }
+        else if (state.state == controllerActions::STAND){
```
We're just checking the winner here instead of based on the check_winner state

orselane commented on main/include/src/atomic_models/controller.hpp:
```diff
@@ -0,0 +1,203 @@
+            const auto lastInput = (decisionInPort->getBag()).back();
+            switch (lastInput) {
+                case decision::HIT:
+                    state.state = controllerActions::HIT; 
```
This line can be done before the switch, by just setting to last input.

orselane replied:
same with the sigma if they're the same

ArthurAtangana replied:
no as they are a different type

orselane commented on main/include/src/atomic_models/controller.hpp:
```diff
@@ -0,0 +1,203 @@
+            state.current_player = Players::CHALLENGER;
+            state.dealer_score = 0;
+            state.challenger_score = 0;
+            state.state = controllerActions::IDLE;
```
check for repition

orselane commented on main/include/src/atomic_models/controller.hpp:
```diff
@@ -0,0 +1,203 @@
+        case controllerActions::START: out << "START"; break;
+        case controllerActions::HIT: out << "HIT"; break;
+        case controllerActions::STAND: out << "STAND"; break;
+        case controllerActions::CHECK_WINNER: out << "CHECK_WINNER"; break;
```
remove check_winner

orselane commented:
Approved, assuming you address the comments later.

ArthurAtangana committed:
Add ports in controller

- Controller command out was sending DRAW_CHALLENGER or DRAW_DEALER to the
deck:
	* Rework the commands to be just SHUFFLE or DRAW, and add a port for the player when DRAW.
- Remove CHECK_WINNER State as it was irrelevent. specs to follow
- Rename card port to hand port (as controller is receiving the hand
  value and not the card value).

ArthurAtangana committed:
Add ports in deck for the card and players

- fix commands from DRAW_CHALLENGER and DRAW_DEALER to DRAW
- other quality of life fixes
- fix timing for mock game

ArthurAtangana committed:
Repair deck.hpp from new commands (DRAW) and ports

- cleanup main.cpp with duplicate or namechanged runs
